src/webgpu/api/validation/error_scope.spec.ts

```diff
@@ -74,99 +93,187 @@ class F extends Fixture {
+      const uncapturedErrorEvent = await t.expectUncapturedError(() => {
+        t.generateError(errorType);
+      });
+      t.expect(t.isInstanceOfError(errorType, uncapturedErrorEvent.error));
+
+      const error = await t.device.popErrorScope();
+      t.expect(error === null);
+    } else {
+      // Same error as filter
+      t.generateError(errorType);
+      const error = await t.device.popErrorScope();
+      t.expect(t.isInstanceOfError(errorType, error));
+    }
   });
-  t.expect(uncapturedErrorEvent.error instanceof GPUValidationError);
 
-  const error = await t.device.popErrorScope();
-  t.expect(error === null);
-});
-
-g.test('push,popping_sibling_error_scopes_must_be_balanced').fn(async t => {
-  {
+g.test('empty')
+  .desc(
+    `
+Tests that popping an empty error scope stack should reject.
+    `
+  )
+  .fn(async t => {
     const promise = t.device.popErrorScope();
     t.shouldReject('OperationError', promise);
-  }
+  });
 
-  const promises = [];
-  for (let i = 0; i < 1000; i++) {
-    t.device.pushErrorScope('validation');
-    promises.push(t.device.popErrorScope());
-  }
-  const errors = await Promise.all(promises);
-  t.expect(errors.every(e => e === null));
+g.test('parent_scope')
+  .desc(
+    `
+Tests that an error bubbles to the correct parent scope.
+
+- Different error types as the parent scope
+- Different depths of non-capturing filters for the generated error
+    `
+  )
+  .params(u =>
+    u.combine('errorFilter', kErrorScopeFilters).combine('stackDepth', [1, 10, 100, 1000, 100000])
+  )
+  .fn(async t => {
+    const { errorFilter, stackDepth } = t.params;
+    t.device.pushErrorScope(errorFilter);
+
+    // Push a bunch of error filters onto the stack (none that match errorFilter)
+    const unmatchedFilters = kErrorScopeFilters.filter(filter => {
+      return filter !== errorFilter;
+    });
+    for (let i = 0; i < stackDepth; i++) {
+      t.device.pushErrorScope(unmatchedFilters[i % unmatchedFilters.length]);
+    }
+
+    // Cause the error and then pop all the unrelated filters.
+    t.generateError(errorFilter);
+    const promises = [];
+    for (let i = 0; i < stackDepth; i++) {
+      promises.push(t.device.popErrorScope());
+    }
+    const errors = await Promise.all(promises);
+    t.expect(errors.every(e => e === null));
+
+    // Finally the actual error should have been caught by the parent scope.
+    const error = await t.device.popErrorScope();
+    t.expect(t.isInstanceOfError(errorFilter, error));
+  });
 
-  {
-    const promise = t.device.popErrorScope();
-    t.shouldReject('OperationError', promise);
-  }
-});
+g.test('current_scope')
+  .desc(
+    `
+Tests that an error does not bubbles to parent scopes when local scope matches.
+
+- Different error types as the current scope
+- Different depths of non-capturing filters for the generated error
+    `
+  )
+  .params(u =>
+    u.combine('errorFilter', kErrorScopeFilters).combine('stackDepth', [1, 10, 100, 1000, 100000])
```

Comment:
I don't think a stackDepth of 100k is needed on every test, since the main point of that is to make sure that (1) pushing that many scopes, and (2) generating an error with that many scopes on the stack, don't crash anything. Maybe keep it just on this test and reduce the others to 1000? Just to reduce the test time.

Reply:
Done.